[WIP] Refactor NavigationMapRoute

danesfeder opened the pull request:
Closes #704 

This PR aims to refactor the monstrous `NavigationMapRoute` and solve the issues of the route freezing the UI thread when in the process of drawing new routes.  

Part 1 is processing traffic data on a background thread.  Part 2 is going to be optimizing how we update the map source after we have the data we need.

danesfeder commented:
@tobrun @LukasPaczos 

Processing the traffic data has seemed to help to UI thread blockage by a decent amount, there is still some blockage but most likely by [Part 2](#928):

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/8434572/46434402-82f8ed00-c721-11e8-9fa0-8516a319c9af.gif)

If either of you have a minute to check this code out, after my changes,  I'm wondering if there's anything glaring causing this behavior I'm seeing:

![ezgif com-video-to-gif 1](https://user-images.githubusercontent.com/8434572/46434432-9c019e00-c721-11e8-8548-14f257de8b4e.gif)

It seems the previous route geometry is being drawn before the new updated geometry is drawn.  I'm clearing all of the data for the feature collections, so I've been having a tough time tracking this one down.  The `NavigationMapRoute` still needs some cleanup for #928, so hopefully the bug above will iron itself out with a bit more organization.

danesfeder commented:
Update here - had an awesome pairing session with @LukasPaczos and we were able to identify some great areas of optimization.

> It seems the previous route geometry is being drawn before the new updated geometry is drawn. I'm clearing all of the data for the feature collections, so I've been having a tough time tracking this one down.

Also were able to track this down ^ thanks for the eyes on that @LukasPaczos!

LukasPaczos commented:
> My main struggles were around handling the alternative route selection (differentiating between two alternative route layers in the same alternative Source). I opted to go with the previous setup, but refactored a bit.

In order to achieve that using only one source, you'll need to add a property to each "CongestionFeature" that indicates, that this feature is a part of a selected, or an alternative route. Then, you need to verify that in your layer setup.
```
Feature feature = Feature.fromGeometry(congestionLineString);
...
feature.addBooleanProperty(PRIMARY_ROUTE_PROPERTY, isPrimary);
```
```
PropertyFactory.lineColor(
  switchCase(
    get(PRIMARY_ROUTE_PROPERTY), match(
      Expression.toString(get(RouteConstants.CONGESTION_KEY)),
      color(routeDefaultColor),
      stop("moderate", color(routeModerateColor)),
      stop("heavy", color(routeSevereColor)),
      stop("severe", color(routeSevereColor))
    ),
    match(
      Expression.toString(get(RouteConstants.CONGESTION_KEY)),
      color(alternativeRouteDefaultColor),
      stop("moderate", color(alternativeRouteModerateColor)),
      stop("heavy", color(alternativeRouteSevereColor)),
      stop("severe", color(alternativeRouteSevereColor))
    )
  )
)
```

Haven't tested it, but I believe setting the layer up this way, would let you add it to the map just once instead of re-adding it with each route selection, because you are only using data-driven values for the segments' colors. All you need to do, is reset the `PRIMARY_ROUTE_PROPERTY` property and update the source.

> The fix is we need to notify the GeoJsonSource that we removed the layers right? I hit a snag trying to figure this out. How do we "refresh" the source when there is no data to refresh it with?

If coupled with the above setup, there would be no need for removing the layer! But besides that, have you tried reseting the source with an empty collection once `clearRouteListData` is called?

LukasPaczos commented:
Continuing slack convo, I'm unable to reproduce the issue. Would you mind pushing your current setup that creates the issue? Is there a problem with other `LineLayer`s (eg `routeModerateColor`)?

danesfeder commented on libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/route/MapRouteLine.java:
```diff
@@ -0,0 +1,513 @@
+    TypedArray typedArray = context.obtainStyledAttributes(styleRes, R.styleable.NavigationMapRoute);
+
+    // Primary Route attributes
+    routeDefaultColor = typedArray.getColor(R.styleable.NavigationMapRoute_routeColor,
```
@LukasPaczos this should be returning a `@ColorInt` that can be used with the expression right?

danesfeder commented:
@LukasPaczos I was actually able to narrow down the color issue to using `lineColor` for the scale expressions 🙈 I updated to `lineWidth` and I think this is the last issue:

```
E/Mbgl: {igation.testapp}[JNI]: Error setting property: line-width Only one zoom-based "step" or "interpolate" subexpression may be used in an expression.
```

Expression:

```
      lineWidth(
        switchCase(
          get("primary-route"), interpolate(
            exponential(1.5f), zoom(),
            stop(10f, 7f),
            stop(14f, 10.5f * routeScale),
            stop(16.5f, 15.5f * routeScale),
            stop(19f, 24f * routeScale),
            stop(22f, 29f * routeScale)
          ),
          interpolate(
            exponential(1.5f), zoom(),
            stop(10f, 7f),
            stop(14f, 10.5f * alternativeRouteScale),
            stop(16.5f, 15.5f * alternativeRouteScale),
            stop(19f, 24f * alternativeRouteScale),
            stop(22f, 29f * alternativeRouteScale)
          )
        )
      )
```

I think we may be missing the `match` seen in the color switch expressions?  I've pushed the latest code if you have a cycle to check it out.  Thanks for all the help 🙏

LukasPaczos commented:
The unfortunate limitation of expressions is that the `zoom` expression can only be a part of the topmost expression. In this case, we'd probably need to rework the expression to something along the lines of:
```
lineWidth(
  interpolate(
    exponential(1.5f), zoom(),
    stop(10f, 7f),
    stop(14f, product(literal(10.5f),
      switchCase(
        get("primary-route"), literal(routeScale),
        literal(alternativeRouteScale)))),
    stop(16.5f, product(literal(15.5f),
      switchCase(
        get("primary-route"), literal(routeScale),
        literal(alternativeRouteScale)))),
    stop(19f, product(literal(24f),
      switchCase(
        get("primary-route"), literal(routeScale),
        literal(alternativeRouteScale)))),
    stop(22f, product(literal(29f),
      switchCase(
        get("primary-route"), literal(routeScale),
        literal(alternativeRouteScale))))
  )
)
```

danesfeder commented:
@LukasPaczos it's alive and working!  Thanks so much for the help with all of this.  

Still going to add some more tests and would like to do some performance testing to compare against other cross-country routes. 

The last blocker is the API to hide / show alternative routes.  Before this was done by hiding / showing alternative `Layer`s.  Now that primary and alternative routes live on the same layer, do you have a suggestion on how to hide / show them.  I was thinking maybe another case in the expression for `lineColor` where we basically set the color to an invisible hex?

LukasPaczos commented:
I think in this case filtering might come in handy. We can filter visible/hidden routes based on the `PRIMARY_ROUTE_PROPERTY_KEY`. For ex., if you'd like only the primary route to be visible, you can use:
```
layer.setFilter(Expression.eq(Expression.get(PRIMARY_ROUTE_PROPERTY_KEY), true))
```

danesfeder commented:
@LukasPaczos @Guardiola31337 @devotaaabel okay this is ready for review 💯 

Lukas, my final concern is an intermittent native crash I'm seeing in CI instrumentation tests.  If you can 👀 any bad practices with the Maps SDK APIs in this code, please point them out and we can fix them up.  Thanks again for all of your help bringing this to the finish line ❤️

danesfeder commented:
@LukasPaczos I know this may not be helpful, but here is a test matrix producing the native crash I'm seeing https://console.firebase.google.com/u/0/project/mapbox-navigation-android/testlab/histories/bh.5fbde16d4049dec/matrices/6053787514803369453/executions/bs.2a03040076e72c35

danesfeder commented:
Hey @LukasPaczos 👋 in https://github.com/mapbox/mapbox-navigation-android/pull/1387/commits/ef6686c135cd4da4cabb3d63122f21a60f56a149 I updated the `MapRouteLine` and `MapRouteArrow` to each have an `onDestroy` that aims to clean up the layers.   I also updated the arrow code to remove the layers if we find they exist.  

I did find that I'm able to reliably reproduce a native crash in `NavigationMapRouteActivity` and it looks to be related to these changes.  Would you mind having a look at this and seeing where I went wrong (this does not occur in `master`)?  Thank you!

Test app flow to repro (on final screen click the FAB):

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/8434572/49477255-5a0cdb00-f7ea-11e8-839e-135ea09af242.gif)

danesfeder commented:
@LukasPaczos awesome thanks!  It's working now:

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/8434572/49598375-7a54ab00-f94c-11e8-83e5-7f5bdab06461.gif)

Still seeing `null pointer deference` in firebase though 😞 could this be related to removing all of the layers in `onDestroy` but not the `GeoJsonSource`s?

danesfeder commented:
Yeah that seems to have fixed it, thanks @tobrun

danesfeder commented:
This is ready for review!

danesfeder commented on libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/route/MapRouteClickListener.java:
```diff
@@ -0,0 +1,88 @@
+    return routeLineStrings == null || routeLineStrings.isEmpty() || !alternativesVisible;
+  }
+
+  private void findClickedRoute(@NonNull LatLng point, HashMap<LineString, DirectionsRoute> routeLineStrings,
```
This sounds great - let's ticket this out, while keeping in mind using this implementation with a bounding box setup to increase the click region.

danesfeder replied:
ticketed in https://github.com/mapbox/mapbox-navigation-android/issues/1634

danesfeder commented:
@LukasPaczos this is up-to-date now with your changes, thanks again!  My only remaining question is https://github.com/mapbox/mapbox-navigation-android/pull/1387#discussion_r241908587

devotaaabel commented on libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/route/FeatureProcessingTask.java:
```diff
@@ -0,0 +1,95 @@
+  @Override
+  protected void onPostExecute(Void result) {
+    super.onPostExecute(result);
+    Runtime.getRuntime().gc();
```
Why is it manually garbage collecting?

Guardiola31337 commented on libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/route/NavigationMapRoute.java:
```diff
@@ -348,627 +263,81 @@ public void setOnRouteSelectionChangeListener(
-      MapUtils.addLayerToMap(mapboxMap, waypointLayer, belowLayer);
-    }
+  @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
+  public void onDestroy() {
```
I believe this was added to workaround 👇 commented in https://github.com/mapbox/mapbox-navigation-android/pull/1387#issuecomment-444284513
> I updated the `MapRouteLine` and `MapRouteArrow` to each have an `onDestroy` that aims to clean up the layers. I also updated the arrow code to remove the layers if we find they exist.

Question (especially for @LukasPaczos) - Shouldn't this be handled gracefully from the Maps SDK side when calling `MapboxMap#onDestroy`? Have we explored this option? I can cut a ticket in `mapbox-gl-native` to discuss further if you think this should be addressed upstream (if there isn't one already 😅). Does that make sense? Let me know what you think.

LukasPaczos replied:
Correct, there is no need for clearing the layers on the actual `MapView` destroy. You can even see logs like
```
E/Mbgl-NativeMapView: You're calling `removeLayer` after the `MapView` was destroyed, were you invoking it after `onDestroy()`?
```
because it's called after the call to `MapView#onDestroy` and it doesn't have any real impact.

I believe that the initial idea here was to clear layers when the `NavigationMapRoute` is detached from the map so that the navigation SDK doesn't leave unused layers on the map that can still be in use otherwise. If a solution for this minor use-case seems not so relevant in the grand scheme of things, it can definitely be tackled by a separate PR.

Guardiola31337 replied:
Sorry if I didn't make myself clear. I was wondering if the Maps SDK should be in charge of the "layers cleanup" somehow - ensuring that any layers added are removed properly when in `onDestroy`. Probably it has more implications / considerations that I'm not taking into account / missing but IMO this should be maps responsibility - obviously calling `removeLayer` afterwards is a dev implementation mistake (even in this case which doesn't seem to affect - other than the logs printed out). Is that what you're saying? What am I missing? I'm confused 😅 

> I believe that the initial idea here was to clear layers when the NavigationMapRoute is detached from the map so that the navigation SDK doesn't leave unused layers on the map that can still be in use otherwise.

Not sure if I follow, I think ☝️ was the part that make me 😕 Could you clarify?

> If a solution for this minor use-case seems not so relevant in the grand scheme of things, it can definitely be tackled by a separate PR.

Sure thing! This was a question 🤔 that I wanted to fully understand. Not a blocker at all, I agree.

LukasPaczos replied:
> I was wondering if the Maps SDK should be in charge of the "layers cleanup" somehow - ensuring that any layers added are removed properly when in onDestroy.

This is already the case. There is no need for a manual layer cleanup when the map is destroyed, that's what I meant above.

> Not sure if I follow, I think ☝️ was the part that make me 😕 Could you clarify?

I was going with an impression that the `NavigationMapRoute` is a standalone object that can be added and removed from the map. If that's not the case, there's nothing to worry about :)

LukasPaczos replied:
Yes, if the map will never outlive `MapRouteLine`, we shouldn't worry about removing those layers.

Guardiola31337 replied:
Got it! Thanks @LukasPaczos 🙏

cc'ing @danesfeder to double check / confirm and remove.

LukasPaczos commented on libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/route/FeatureProcessingTask.java:
```diff
@@ -0,0 +1,90 @@
+    callback.onRouteFeaturesProcessed(routeFeatureCollections, routeLineStrings);
+  }
+
+  private FeatureCollection createRouteFeatureCollection(DirectionsRoute route, boolean isPrimary) {
```
I think you are right, not sure what I've seen here before... 🤔

LukasPaczos commented:
Thanks for addressing previous comments!

I found another 2 minor issues:
- The previous routes are being displayed after the style load (even though they were previously hidden).
![ezgif com-video-to-gif 100](https://user-images.githubusercontent.com/16925074/50161014-e800bf80-02da-11e9-95d2-62b7e3aa648b.gif)
- _Sometimes_ the alternative routes are not displayed and show up only after the style reload. This isn't reproducible on demand, but every 5 route generations or so the issues shows itself. Also, a minor issue, that a selected route is not saved across style changes.
![ezgif com-video-to-gif 8](https://user-images.githubusercontent.com/16925074/50161123-2dbd8800-02db-11e9-9683-0d5e111a0fc4.gif)

danesfeder commented:
@LukasPaczos great catches!  I believe I was able to address both of them with this latest round of updates.  What was happening is we were trying to recreate these options (visibility, primary route, etc.) before the feature processing task was finished.  

Previous route not being displayed / primary route selection maintained

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/8434572/50258848-8c642e00-03cf-11e9-893d-d0dcc7497d7b.gif)

![ezgif com-video-to-gif 1](https://user-images.githubusercontent.com/8434572/50258860-984ff000-03cf-11e9-8291-c25d3cd7556c.gif)

LukasPaczos commented:
LGTM, thanks for the fantastic work @danesfeder!

Leaving to final review to the nav folks.
